src/sparse_blas/backends/mkl_common/mkl_helper.hpp

```diff
@@ -1,56 +1,102 @@
+*  limitations under the License.
 *
-*
-* SPDX-License-Identifier: Apache-2.0
-*******************************************************************************/
+**************************************************************************/
 
-// MKLCPU and MKLGPU backends include
-// This include defines its own oneapi::mkl::sparse namespace with some of the types that are used here: matrix_handle_t, index_base, transpose, uolo, diag.
-#include <oneapi/mkl/spblas.hpp>
+#ifndef _ONEMKL_SRC_SPARSE_BLAS_BACKENDS_MKL_COMMON_MKL_HELPER_HPP_
+#define _ONEMKL_SRC_SPARSE_BLAS_BACKENDS_MKL_COMMON_MKL_HELPER_HPP_
 
-// Includes are set up so that oneapi::mkl::sparse namespace refers to the MKLCPU and MKLGPU backends namespace (oneMKL product)
-// in this file.
-// oneapi::mkl::sparse::detail namespace refers to the oneMKL interface namespace.
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
 
+#include "oneapi/mkl/exceptions.hpp"
 #include "oneapi/mkl/sparse_blas/detail/helper_types.hpp"
 
+#include "sparse_blas/enum_data_types.hpp"
+#include "sparse_blas/macros.hpp"
+
 namespace oneapi::mkl::sparse::detail {
 
-inline auto get_handle(detail::matrix_handle **handle) {
-    return reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t *>(handle);
+/// Return whether a pointer is accessible on the host
+template <typename T>
+inline bool is_ptr_accessible_on_host(sycl::queue &queue, const T *host_or_device_ptr) {
+    auto alloc_type = sycl::get_pointer_type(host_or_device_ptr, queue.get_context());
+    // Note sycl::usm::alloc::host may not be accessible on the host according to SYCL specification.
+    // sycl::usm::alloc::unknown is returned if the pointer is not a USM allocation which is assumed to be a normal host pointer.
+    return alloc_type == sycl::usm::alloc::shared || alloc_type == sycl::usm::alloc::unknown;
+}
+
+/// Throw an exception if the scalar is not accessible in the host
+template <typename T>
+void check_ptr_is_host_accessible(const std::string &function_name, const std::string &scalar_name,
+                                  sycl::queue &queue, const T *host_or_device_ptr) {
+    if (!is_ptr_accessible_on_host(queue, host_or_device_ptr)) {
+        throw mkl::invalid_argument(
+            "sparse_blas", function_name,
+            "Scalar " + scalar_name + " must be accessible on the host for buffer functions.");
+    }
 }
 
-inline auto get_handle(detail::matrix_handle *handle) {
-    return reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t>(handle);
+/// Return a scalar on the host from a pointer to host or device memory
+/// Used for USM functions
+template <typename T>
+inline T get_scalar(sycl::queue &queue, const T *host_or_device_ptr) {
+    if (is_ptr_accessible_on_host(queue, host_or_device_ptr)) {
+        return *host_or_device_ptr;
+    }
+    T scalar;
+    auto event = queue.copy(host_or_device_ptr, &scalar, 1);
+    event.wait_and_throw();
```

Comment:
I suppose this function is for use in `alpha`/`beta` arguments of APIs? Is this specifically used _only_ in case of `sycl::buffer` APIs, because I thought the whole point of providing `void * {alpha, beta}` to the APIs was to allow asynchronous calculation of `alpha`/`beta` from some previous APIs in USM case?

Reply:
Yes this is for the `alpha` and `beta` parameters. In fact while working on the cuSPARSE backend I realized I forgot to use this function for the Intel oneMKL backends. I have added "support" for these scalars to be USM pointers on the device and tests in https://github.com/oneapi-src/oneMKL/pull/500/commits/d4528466da3aa1608ddb2d60c5c8bff88306a7c4 and a note in the documentation in https://github.com/oneapi-src/oneMKL/pull/500/commits/002d78867f238d0e357b1cd8bdd63bc6554a6463.
The Intel oneMKL backends need these parameters on the host so I was thinking of copying the parameter to the host if needed even if this adds a synchronization point. Alternatively we could throw unsupported.
We won't need to perform the copy for the cuSPARSE backend.
We agreed in the spec that these scalars must be on the host for the buffer API.